Improve support for 64-bit types on Android.

TechnoPorg opened the pull request:
Fixes #67193.
This update mappings in the JNI functions to better support 64-bit integers, and adds support for 64-bit float arrays.
The code is mostly reused from 32-bit types.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

Commit:
Improve support for 64-bit types on Android.

This update mappings in the JNI functions to better support 64-bit integers, and adds support for 64-bit float arrays.
The code is mostly reused from 32-bit types.

m4gr3d commented on platform/android/jni_utils.cpp:
```diff
@@ -167,8 +167,14 @@ jvalret _variant_to_jvalue(JNIEnv *env, Variant::Type p_type, const Variant *p_a
 		} break;
-
-			// TODO: This is missing 64 bits arrays, I have no idea how to do it in JNI.
+		case Variant::PACKED_FLOAT64_ARRAY: {
+			Vector<double> array = *p_arg;
+			jdoubleArray arr = env->NewDoubleArray(array.size());
+			const double *r = array.ptr();
+			env->SetDoubleArrayRegion(arr, 0, array.size(), r);
+			v.val.l = arr;
+			v.obj = arr;
+		} break;
```
I think you also need to update the mappings in `get_jni_type(...)` and `get_jni_sig(...)`

m4gr3d commented on platform/android/api/jni_singleton.h:
```diff
@@ -149,9 +149,18 @@ class JNISingleton : public Object {
 				ret = sarr;
 				env->DeleteLocalRef(arr);
 			} break;
+			case Variant::PACKED_FLOAT64_ARRAY: {
```
For completion, we can also add the implementation for the `PACKED_INT64_ARRAY` variant type, using `long` instead of `int`.

TechnoPorg commented:
I have now added support for 64-bit integer arrays and updated `get_jni_type` and `get_jni_sig`.
I've also added some lines that should allow for more flexible conversions involving long integers.
I really appreciate your reviewing this, because I unfortunately don't have access to an Android device, so I'm basically going blind 🙂.

TechnoPorg commented:
Rebased this PR following #67581, which contained some of the changes included here.

TechnoPorg commented:
@m4gr3d Sorry to bump, but would you mind taking a look at this when you have time?

m4gr3d reviewed (approved):
Looks good!

akien-mga commented:
Thanks!
